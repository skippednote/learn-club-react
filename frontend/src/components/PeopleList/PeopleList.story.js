import React from 'react';
import { storiesOf } from '@storybook/react';
import PeopleList from './index';
import fixture from '../../pages/People/People.fixture';

storiesOf('PeopleList', module).add('Default', () => (
  <PeopleList people={fixture} />
));
