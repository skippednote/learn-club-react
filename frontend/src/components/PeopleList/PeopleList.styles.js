import styled from 'react-emotion';

export const PeopleItem = styled('li')`
  width: 300px;
  text-align: left;
  display: flex;
  justify-content: space-between;
  margin: auto;
  line-height: 42px;
  font-size: 14px;
  border-bottom: 1px solid #ddd;
`;
PeopleItem.displayName = 'PeopleItem';
