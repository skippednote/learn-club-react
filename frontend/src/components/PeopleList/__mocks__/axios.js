const people = [
  {
    id: 0,
    name: 'Ankur',
    role: 'CEO'
  },
  {
    id: 1,
    name: 'Aliya',
    role: 'Developer'
  },
  {
    id: 2,
    name: 'Bassam',
    role: 'Developer'
  }
];

module.exports = {
  get: jest.fn(() => Promise.resolve({ data: people }))
};
