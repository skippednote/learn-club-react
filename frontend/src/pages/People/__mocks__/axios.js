const fixture = [
  {
    id: 0,
    name: 'Ankur',
    role: 'CEO'
  },
  {
    id: 1,
    name: 'Aliya',
    role: 'Developer'
  },
  {
    id: 2,
    name: 'Bassam',
    role: 'Developer'
  }
];

module.exports = {
  fixture,
  get: jest.fn(() => Promise.resolve({ data: fixture }))
};
