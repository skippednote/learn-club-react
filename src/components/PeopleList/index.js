// @flow
import React, { Component } from 'react';
import { List, Item } from './PeopleList.styles';
import type { Person } from '../../pages/People/People.types';

type Props = {
  people: Array<Person>
};

class PeopleList extends Component<Props> {
  render() {
    const { people } = this.props;
    return (
      <List className="peoplelist">
        {people.map(({ id, name, role }) => (
          <Item key={id}>
            <span>{name}</span>
            <span>{role}</span>
          </Item>
        ))}
      </List>
    );
  }
}

export default PeopleList;
