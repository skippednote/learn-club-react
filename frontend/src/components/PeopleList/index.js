// @flow
import React, { Component } from 'react';
import axios from 'axios';
import { PeopleItem } from './PeopleList.styles';

type Person = {
  id: number,
  name: string,
  role: string
};

type State = {
  people: Array<Person>,
  error: string | null,
  loading: boolean
};

class PeopleList extends Component<{}, State> {
  state: State = {
    people: [],
    error: null,
    loading: false
  };

  async componentDidMount() {
    this.setState({ loading: true });

    try {
      const { data: people }: { data: Array<Person> } = await axios.get(
        '/people.json'
      );
      this.setState({ people, loading: false });
    } catch (error) {
      console.log(error);
      this.setState({ error: error, loading: false });
    }
  }

  render() {
    const { people, error, loading } = this.state;
    if (error) {
      return (
        <div className="error">There was an error loading list of People.</div>
      );
    }
    if (loading) {
      return <div className="loading">Loading the list of People...</div>;
    }
    return (
      <ul className="peoplelist">
        {people.map(({ id, name, role }) => (
          <PeopleItem key={id}>
            <span>{name}</span>
            <span>{role}</span>
          </PeopleItem>
        ))}
      </ul>
    );
  }
}

export default PeopleList;
