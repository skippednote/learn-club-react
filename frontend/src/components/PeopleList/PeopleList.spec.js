import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import axios from 'axios';
import PeopleList from './index';

describe('PeopleList', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<PeopleList />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('requests data on #componentDidMount', async () => {
    const wrapper = shallow(<PeopleList />);
    await wrapper.instance().componentDidMount();
    expect(axios.get).toHaveBeenCalled();
    expect(axios.get).toHaveBeenCalledWith('/people.json');
  });

  it('fetches list of person on #componentDidMount', async () => {
    const wrapper = shallow(<PeopleList />);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('people')).toHaveLength(3);
    wrapper.update();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders an error if request fails', async () => {
    const wrapper = shallow(<PeopleList />);
    await wrapper.instance().componentDidMount();
    wrapper.setState({
      error: { message: 'Request failed with status code 404' }
    });
    const error = wrapper.find('.error');
    expect(error.text()).toContain(
      'There was an error loading list of People.'
    );
  });
});
