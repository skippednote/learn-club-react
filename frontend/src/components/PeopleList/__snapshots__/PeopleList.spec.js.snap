// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleList fetches list of person on #componentDidMount 1`] = `
<ul
  className="peoplelist"
>
  <PeopleItem
    key="0"
  >
    <span>
      Ankur
    </span>
    <span>
      CEO
    </span>
  </PeopleItem>
  <PeopleItem
    key="1"
  >
    <span>
      Aliya
    </span>
    <span>
      Developer
    </span>
  </PeopleItem>
  <PeopleItem
    key="2"
  >
    <span>
      Bassam
    </span>
    <span>
      Developer
    </span>
  </PeopleItem>
</ul>
`;

exports[`PeopleList matches snapshot 1`] = `
<div
  className="loading"
>
  Loading the list of People...
</div>
`;
